### PR TITLE
Introducing SenderPolicy to DPL processors

### PR DESCRIPTION
**Pull request by matthiasrichter**
`SenderPolicy` describes on per-device basis when completed objects are sent, either at the end of computation - the default and current behavior - or immediately when ready. Depending on the `SenderPolicy` operation, the `DataProcessingDevice` configures the `MessageContext` with a sender callback. The sender policy can be customized in the usual DPL manner.

`MessageContext` has been extended to create context objects embedded into a scope handler. The context object is scheduled, when the handler goes out of scope in the user code. Currently, the `DataAllocator::snapshot` API method uses this feature.

This PR will be the foundation for refactoring the digitizer workflow and to solve the memory problems with large data sets. Enabling the prompt sending also for `DataAllocator::make` remains to be finalized, but the work is nearly finished. Will be done after refactoring the digitizer workflow. The support also has to be implemented to other contexts, though the `RootObjectContext` is already obsolete and `StringContext` can be easily implemented in the same way, including pmr support.

@ktf, not sure if `SenderPolicy` is the appropriate name. I have implemented in parallel to `CompletionPolicy`. I have considered to combine the two, but in the end I thought its two different things.

### CODE REVIEW

File: Framework/Core/test/test_StaggeringWorkflow.cxx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                         return spec.name.find("processor") == 0;
+                       },
+                       [](auto const&) { return o2::framework::CompletionPolicy::CompletionOp::Consume; } });
+}
+
+#include "Framework/runDataProcessing.h"
+
+using namespace o2::framework;
+
+#define ASSERT_ERROR(condition)                                   \
```
Comment (ktf) [inline Framework/Core/test/test_StaggeringWorkflow.cxx]: We should move ASSERT_ERROR to `Framework/Logger`. I think we have it already in a few places.

Comment (matthiasrichter) [reply Framework/Core/test/test_StaggeringWorkflow.cxx]: true, will check

**Comment by matthiasrichter** [comment]
> I would call it `DispatchPolicy`...

perfect!

As for the matching of specific outputs, this is not implemented so far. I would leave this as a later extension. Right now the dispatch policy applies to all outputs.

**Comment by ktf** [comment]
I meant something different. It's now possible to leave the subSpec unspecified (e.g. via `OutputSpec{"binding", {"TST", "A1"}}`) paired with this it could make parallel processing of chunks of data very easy.

**Comment by matthiasrichter** [comment]
I need have a closer look to the OutputSpec with wildcard. I do touch the matching of an output to a channel, so I think all that is unaffected. Sounds like the combination opens up new possibilities.

I have renamed the new policy and now the code is consistently using `DispatcherPolicy`. Will move the ASSERT_ERROR` definition in a separate PR.

**Comment by ktf** [comment]
Thanks. whenever you are ready we can merge.

**Pull request merged**